Source/WebCore/page/writing-tools/WritingToolsController.mm

```diff
@@ -632,6 +640,42 @@
     m_page->chrome().client().proofreadingSessionUpdateStateForSuggestionWithID(data.sessionID, WritingTools::TextSuggestion::State::Reviewing, data.suggestionID);
 }
 
+void WritingToolsController::respondToUnappliedEditing(EditCommandComposition* composition)
+{
+    for (auto& value : m_states.values()) {
+        auto* state = std::get_if<CompositionState>(&value);
+        if (!state)
+            continue;
+
+        auto appliedCommandIsWritingToolsCommand = std::ranges::any_of(state->reappliedCommands, [&](const auto& command) {
+            return &command->ensureComposition() == composition;
+        });
+
+        if (!appliedCommandIsWritingToolsCommand)
+            continue;
+
+        state->unappliedCommands.append(state->reappliedCommands.takeLast());
+    }
+}
+
+void WritingToolsController::respondToReappliedEditing(EditCommandComposition* composition)
+{
+    for (auto& value : m_states.values()) {
+        auto* state = std::get_if<CompositionState>(&value);
+        if (!state)
+            continue;
+
+        auto appliedCommandIsWritingToolsCommand = std::ranges::any_of(state->unappliedCommands, [&](const auto& command) {
```

Comment:
I think we can make a reusable helper for this?

```
static bool appliedCommandIsWritingToolsCommand(const auto& commands, EditCommandComposition* composition)
{
    return std::ranges::any_of(commands, [composition](const auto& command) {
        return &command->ensureComposition() == composition;
    });
}
```

Reply:
+1 if a method on the object itself is not possible.